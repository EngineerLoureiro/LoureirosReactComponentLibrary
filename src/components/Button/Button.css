.btn {
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 6px;
  flex-grow: 1;
  padding: 0.5rem 1.25rem;
  border-radius: 4px;
  box-shadow: 0px 1px 2px 0 rgb(0 0 0 / 0.06), 0px 1px 3px 0 rgb(0 0 0 / 0.1);
  font-family: "Noto Sans";
  white-space: nowrap;
}

.btn:focus {
  box-shadow: 0px 0px 0px 4px
    rgb(0.2666666805744171 0.2980392277240753 0.9058823585510254 / 0.12);
}
.btn:disabled {
  background-color: #f5f5f5;
  color: #a3a3a3;
}

.btn-primary {
  background-color: #4338ca;
  box-shadow: 0px 1px 2px 0 rgb(0 0 0 / 0.06), 0px 1px 3px 0 rgb(0 0 0 / 0.1);
  color: #ffffff;
}

.btn-primary:hover,
.btn-primary:focus {
  background-color: #3730a3;
}

.btn-secondary {
  background-color: #ffffff;
  border-style: solid;
  border-width: 0.5px;
  border-color: #e5e5e5;
  box-shadow: 0px 1px 2px 0 rgb(0 0 0 / 0.06), 0px 1px 3px 0 rgb(0 0 0 / 0.1);
  color: #171717;
}

.btn-secondary:hover,
.btn-secondary:focus {
  background-color: #fafafa;
  border-width: 1px;
}
