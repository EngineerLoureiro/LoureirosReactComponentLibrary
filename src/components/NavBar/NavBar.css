.navbar {
  display: grid;
  grid-template-columns: auto 1fr auto;
  align-items: center;
  top: 0;
  position: absolute;
  width: 100%;
  padding: 16px clamp(16px, 5vw, 112px) 0px;
  overflow-x: hidden;
}
.navbar__brand {
  display: flex;
  align-items: center;
  gap: 10px;
}

.navbar__links,
.navbar__buttons {
  display: flex;
  gap: 32px;
  list-style: none;
}

.navbar__buttons {
  gap: 16px;
  justify-content: flex-end;
}

.navbar__mobile-menu-icon {
  display: none;
  background: none;
}

.navbar__mobile-menu-icon:focus {
  background-color: #ffffff;
  border-radius: 4px;
  box-shadow: 0px 0px 0px 4px
    rgb(0.2666666805744171 0.2980392277240753 0.9058823585510254 / 0.12);
  outline: none;
}

.navbar__mobile-menu-icon:disabled {
  color: #a3a3a3;
}

h1 {
  font-family: "Noto Sans";
  font-weight: 700;
  font-size: 16px;
  line-height: 24px;
  color: #171717;
  white-space: nowrap;
}

.navbar__slideout-menu_links,
.navbar__slideout-menu_buttons {
  display: flex;
  gap: clamp(14px, 6.4vw, 24px);
  list-style: none;
  flex-direction: column;
  align-items: flex-start;
  padding-left: 0;
  margin: 0;
}
.navbar__slideout-menu_buttons {
  margin-top: auto;
  align-items: center;
  overflow-x: hidden;
  align-items: stretch;
}

@media (max-width: 768px) {
  .navbar {
    display: flex;
    justify-content: space-between;
    padding: 16px 32px;
  }
  .navbar__links,
  .navbar__buttons {
    display: none;
  }

  .navbar__mobile-menu-icon {
    display: block;
  }

  .navbar__slideout-menu_buttons-list_item {
    display: flex;
  }
}

@media (max-width: 375px) {
  .navbar {
    padding: 16px 16px 0px;
  }
}
