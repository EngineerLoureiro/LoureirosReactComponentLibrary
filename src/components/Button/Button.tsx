import classNames from "classnames";
import "./Button.css";

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  label: string;
  backgroundColor?: string;
  textColor?: string;
  variant?: "primary" | "secondary";
}

export function Button({
  label,
  className,
  backgroundColor,
  textColor,
  style,
  onClick,
  variant,
  disabled,
  ...rest
}: ButtonProps) {
  return (
    <button
      className={classNames("btn", variant && `btn-${variant}`, className)}
      style={{ backgroundColor, color: textColor, ...style }}
      onClick={onClick}
      disabled={disabled}
      {...rest}
    >
      {label}
    </button>
  );
}
