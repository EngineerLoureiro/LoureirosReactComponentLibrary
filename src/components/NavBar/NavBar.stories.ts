// Replace your-framework with the framework you are using, e.g. react-vite, nextjs, nextjs-vite, etc.
import type { Meta, StoryObj } from "@storybook/react-vite";
import { BsHouseFill } from "react-icons/bs";

import { NavBar } from "./NavBar";

import * as ButtonLayout from "../Button/Button.stories";
import * as LinkLayout from "../Link/Link.stories";
import { ButtonProps, Button } from "../Button";
import { LinkProps, Link } from "../Link";
import React from "react";

//ðŸ‘‡ This default export determines where your story goes in the story list
const meta = {
  title: "Loureiro'sComponentLibrary/NavBar",
  component: NavBar,
  subcomponents: { Button, Link },
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: "fullscreen",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ["autodocs"],
} satisfies Meta<typeof NavBar>;

export default meta;
type Story = StoryObj<typeof meta>;

const logo: React.ReactNode = React.createElement(BsHouseFill, {
  size: 24,
  color: "blue",
});

export const FirstNavBar: Story = {
  args: {
    title: "Housing Crisis",
    buttons: [ButtonLayout.Primary.args, ButtonLayout.Secondary.args],
    links: [
      LinkLayout.ActiveLink.args,
      LinkLayout.ActiveLink.args,
      LinkLayout.ActiveLink.args,
      LinkLayout.ActiveLink.args,
    ],
    logo: logo,
  },
};
