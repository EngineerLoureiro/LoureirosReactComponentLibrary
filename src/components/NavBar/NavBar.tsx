import { Link, LinkProps } from "../Link";
import { Button, ButtonProps } from "../Button";
import { RxHamburgerMenu } from "react-icons/rx";
import "./NavBar.css";
import { SlideOutMenu } from "../SlideOutMenu";
import { useState } from "react";

export interface NavBarProps {
  logo: React.ReactNode;
  title: string;
  links: LinkProps[];
  buttons: ButtonProps[];
}

export function NavBar({ logo, title, links, buttons }: NavBarProps) {
  const [isMenuOpen, setIsMenuOpen] = useState(false);

  const renderedLinks = links.map((props, index) => {
    return (
      <li key={index}>
        <Link {...props} />
      </li>
    );
  });

  const renderedButtons = buttons.map((props, index) => {
    return (
      <li className="navbar__slideout-menu_buttons-list_item" key={index}>
        <Button {...props} />
      </li>
    );
  });
  return (
    <div className="navbar">
      <div className="navbar__brand" data-testid="navbar-brand">
        {logo}
        <h1>{title}</h1>
      </div>
      <ul className="navbar__links" data-testid="navbar__links">
        {renderedLinks}
      </ul>
      <ul className="navbar__buttons" data-testid="navbar__buttons">
        {renderedButtons}
      </ul>

      <button
        className="navbar__mobile-menu-icon"
        aria-label="Open menu"
        aria-expanded={isMenuOpen}
        onClick={() => setIsMenuOpen(true)}
      >
        <RxHamburgerMenu />
      </button>

      <SlideOutMenu
        isOpen={isMenuOpen}
        onClose={() => setIsMenuOpen(false)}
        logo={logo}
        title={title}
      >
        <ul className="navbar__slideout-menu_links">{renderedLinks}</ul>
        <ul className="navbar__slideout-menu_buttons">{renderedButtons}</ul>
      </SlideOutMenu>
    </div>
  );
}
