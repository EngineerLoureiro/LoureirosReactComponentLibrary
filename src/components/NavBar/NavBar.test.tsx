import React from "react";
import { render, screen, fireEvent, within } from "@testing-library/react";
import { NavBar } from "./NavBar";
import { BsHouseFill } from "react-icons/bs";
import "@testing-library/jest-dom";
import { ButtonProps } from "../Button";

const links = [
  { label: "Home", href: "/" },
  { label: "About", href: "/components/button" },
  {
    label: "Docs",
    href: "/docs",
    target: "_blank",
  },
  { label: "Contact", href: "/components/button", disabled: true },
  {
    label: "GitHub",
    href: "https://github.com/EngineerLoureiro",
    target: "_blank",
  },
];

const buttons: ButtonProps[] = [
  {
    label: "Login",
    variant: "primary",
    onClick: () => alert("Logging in..."),
  },
  {
    label: "Sign Up",
    variant: "secondary",
    onClick: () => alert("Signing up..."),
  },
];

const logo = <BsHouseFill size={24} color="blue" />;

describe("NavBar", () => {
  it("renders logo, title, links and button", () => {
    render(
      <NavBar
        logo={logo}
        title="Housing Crisis"
        links={links}
        buttons={buttons}
      />
    );

    const brand = screen.getByTestId("navbar-brand");
    expect(
      within(brand).getByRole("heading", { name: "Housing Crisis" })
    ).toBeInTheDocument();

    const navBarLinks = screen.getByTestId("navbar__links");
    links.forEach((link) => {
      expect(
        within(navBarLinks).getByRole("link", { name: link.label })
      ).toHaveAttribute("href", link.href);
    });
    const navBarButtons = screen.getByTestId("navbar__buttons");
    buttons.forEach((button) => {
      expect(
        within(navBarButtons).getByRole("button", { name: button.label })
      ).toBeInTheDocument();
    });
  });

  it("opens and closes the mobile menu when hamburguer and close buttons are clicked", () => {
    render(
      <NavBar
        logo={logo}
        title="Housing Crisis"
        links={links}
        buttons={buttons}
      />
    );

    const hamburguerButton = screen.getByRole("button", { name: /open menu/i });
    fireEvent.click(hamburguerButton);

    const closeButton = screen.getByRole("button", { name: /close menu/i });
    expect(closeButton).toBeVisible();

    fireEvent.click(closeButton);
    const slideoutMenu = screen.getByTestId("slideout-menu");
    expect(slideoutMenu).not.toHaveClass("open");
  });
  it("has proper aria attributes", () => {
    render(
      <NavBar
        logo={logo}
        title="Housing Crisis"
        links={links}
        buttons={buttons}
      />
    );

    const toggle = screen.getByRole("button", { name: /open menu/i });
    expect(toggle).toHaveAttribute("aria-expanded", "false");

    fireEvent.click(toggle);
    expect(toggle).toHaveAttribute("aria-expanded", "true");
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <NavBar
        logo={logo}
        title="Housing Crisis"
        links={links}
        buttons={buttons}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
