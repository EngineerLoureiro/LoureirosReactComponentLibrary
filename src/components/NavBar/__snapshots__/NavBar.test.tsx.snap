// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="navbar"
  >
    <div
      class="navbar__brand"
      data-testid="navbar-brand"
    >
      <svg
        color="blue"
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        style="color: blue;"
        viewBox="0 0 16 16"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.707 1.5a1 1 0 0 0-1.414 0L.646 8.146a.5.5 0 0 0 .708.708L8 2.207l6.646 6.647a.5.5 0 0 0 .708-.708L13 5.793V2.5a.5.5 0 0 0-.5-.5h-1a.5.5 0 0 0-.5.5v1.293z"
        />
        <path
          d="m8 3.293 6 6V13.5a1.5 1.5 0 0 1-1.5 1.5h-9A1.5 1.5 0 0 1 2 13.5V9.293z"
        />
      </svg>
      <h1>
        Housing Crisis
      </h1>
    </div>
    <ul
      class="navbar__links"
      data-testid="navbar__links"
    >
      <li>
        <a
          class="link"
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          class="link"
          href="/components/button"
        >
          About
        </a>
      </li>
      <li>
        <a
          class="link"
          href="/docs"
          rel="noopener noreferrer"
          target="_blank"
        >
          Docs
        </a>
      </li>
      <li>
        <a
          aria-disabled="true"
          class="link link--disabled"
          href="/components/button"
          tabindex="-1"
        >
          Contact
        </a>
      </li>
      <li>
        <a
          class="link"
          href="https://github.com/EngineerLoureiro"
          rel="noopener noreferrer"
          target="_blank"
        >
          GitHub
        </a>
      </li>
    </ul>
    <ul
      class="navbar__buttons"
      data-testid="navbar__buttons"
    >
      <li
        class="navbar__slideout-menu_buttons-list_item"
      >
        <button
          class="btn btn-primary"
        >
          Login
        </button>
      </li>
      <li
        class="navbar__slideout-menu_buttons-list_item"
      >
        <button
          class="btn btn-secondary"
        >
          Sign Up
        </button>
      </li>
    </ul>
    <button
      aria-expanded="false"
      aria-label="Open menu"
      class="navbar__mobile-menu-icon"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 15 15"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M1.5 3C1.22386 3 1 3.22386 1 3.5C1 3.77614 1.22386 4 1.5 4H13.5C13.7761 4 14 3.77614 14 3.5C14 3.22386 13.7761 3 13.5 3H1.5ZM1 7.5C1 7.22386 1.22386 7 1.5 7H13.5C13.7761 7 14 7.22386 14 7.5C14 7.77614 13.7761 8 13.5 8H1.5C1.22386 8 1 7.77614 1 7.5ZM1 11.5C1 11.2239 1.22386 11 1.5 11H13.5C13.7761 11 14 11.2239 14 11.5C14 11.7761 13.7761 12 13.5 12H1.5C1.22386 12 1 11.7761 1 11.5Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <aside
      class="slideout-menu"
      data-testid="slideout-menu"
    >
      <div
        class="slideout-menu__header"
      >
        <div
          class="slideout-menu__brand"
        >
          <svg
            color="blue"
            fill="currentColor"
            height="24"
            stroke="currentColor"
            stroke-width="0"
            style="color: blue;"
            viewBox="0 0 16 16"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.707 1.5a1 1 0 0 0-1.414 0L.646 8.146a.5.5 0 0 0 .708.708L8 2.207l6.646 6.647a.5.5 0 0 0 .708-.708L13 5.793V2.5a.5.5 0 0 0-.5-.5h-1a.5.5 0 0 0-.5.5v1.293z"
            />
            <path
              d="m8 3.293 6 6V13.5a1.5 1.5 0 0 1-1.5 1.5h-9A1.5 1.5 0 0 1 2 13.5V9.293z"
            />
          </svg>
           
          <h2>
            Housing Crisis
          </h2>
        </div>
        <button
          aria-label="Close menu"
          class="slideout-menu__close"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.72 5.72a.75.75 0 0 1 1.06 0L12 10.94l5.22-5.22a.749.749 0 0 1 1.275.326.749.749 0 0 1-.215.734L13.06 12l5.22 5.22a.749.749 0 0 1-.326 1.275.749.749 0 0 1-.734-.215L12 13.06l-5.22 5.22a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042L10.94 12 5.72 6.78a.75.75 0 0 1 0-1.06Z"
            />
          </svg>
        </button>
      </div>
      <ul
        class="navbar__slideout-menu_links"
      >
        <li>
          <a
            class="link"
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            class="link"
            href="/components/button"
          >
            About
          </a>
        </li>
        <li>
          <a
            class="link"
            href="/docs"
            rel="noopener noreferrer"
            target="_blank"
          >
            Docs
          </a>
        </li>
        <li>
          <a
            aria-disabled="true"
            class="link link--disabled"
            href="/components/button"
            tabindex="-1"
          >
            Contact
          </a>
        </li>
        <li>
          <a
            class="link"
            href="https://github.com/EngineerLoureiro"
            rel="noopener noreferrer"
            target="_blank"
          >
            GitHub
          </a>
        </li>
      </ul>
      <ul
        class="navbar__slideout-menu_buttons"
      >
        <li
          class="navbar__slideout-menu_buttons-list_item"
        >
          <button
            class="btn btn-primary"
          >
            Login
          </button>
        </li>
        <li
          class="navbar__slideout-menu_buttons-list_item"
        >
          <button
            class="btn btn-secondary"
          >
            Sign Up
          </button>
        </li>
      </ul>
    </aside>
  </div>
</DocumentFragment>
`;
